README.md
A feature clustering algorithm for non-targeted mass spectrometric metabolomics data.

To Install from R console:


install.packages("devtools", repos="http://cran.us.r-project.org", dependencies=TRUE)

library(devtools)

install_github("cbroeckl/RAMClustR", build_vignettes = TRUE, dependencies = TRUE)

library(RAMClustR)

vignette("RAMClustR")

RAMClustR publication: RAMClust: A Novel Feature Clustering Method Enables Spectral-Matching-Based Annotation for Metabolomics Data C. D. Broeckling, F. A. Afsar, S. Neumann , A. Ben-Hur, and J. E. Prenni http://pubs.acs.org/doi/abs/10.1021/ac501530d 



ChangeLog
2021-08-11 Corey Broeckling v1.2.2
  * fixed clri bug in write.msp
  * fixed annotation bug - no structures from MSFinder were being assigned
  * fixed a neutral loss format error used for findMain

2021-04-19 Corey Broeckling v1.2.1
  * removed closeAllConnections() call from rc.cmpd.get.pubchem for CRAN compatibility

2021-04-19 Corey Broeckling v1.2.0
  * added pubchem retreival functionality
  * added blank filter to remove background features
  * added normalization to QC samples

2020-04-14 Corey Broeckling v1.1.0
  * splitting ramclustR function into several smaller functions
  * adjust $history to a list.  

2019-09-17 Corey Broeckling v1.0.8
  * remove dependency on XCMS, replace with XCMS as suggestion

2019-08-27 Corey Broeckling v1.0.8
  * add reporting, fix processing history narrative issues. 

2019-04-23 Corey Broeckling v1.0.7
  * trying to improve stability of getClassyFire - bug fix in tryCatch part of problem. 

2019-04-23 Corey Broeckling v1.0.6
  * adapted getClassyFire to limit number of new structure ClassyFications to 5/minute per server regulations
  * bug fix in annotate, get.synonyms

2019-02-15 Corey Broeckling v1.0.5
  * fixed bug in do.findmain

2019-02-12 Corey Broeckling v1.0.4
  * fixed bug in batch.qc correction for MSe data

2019-01-30 Corey Broeckling v1.0.3
  * added write.methods function to export processing history

2019-01-30 Corey Broeckling v1.0.2
  * changed behavior of msfinder import and annotation. Structure results from a relevent database can be used to set the formula, rather than formula being independently assigned.  
  * added rt.only.low.n option to control behavior of clustering with low n (< 5) 

2019-01-18 Corey Broeckling v1.0.1
  * fixed bug in ramclustR function when fewer than five samples. 
  * added function to convert foodb download to MSFinder structure database import format. 

2019-01-17 Corey Broeckling
  * update R, CRAN check, 1.0.0

2019-01-11 Corey Broeckling 
  * check() returns zero errors or warnings, two notes.
  * check.windows() returns BioC warning, unclear how to fix (no reference to BioC apparent)
  * updated do.findmain, altered ramclustR scoring algorithm, made ramclustR default, and changed selection criterion for find.main vs ramclustR scoring when mass assignment is conflicted between the two.  

2018-06-26 Corey Broeckling 
  * check() returns zero errors, warnings, or notes. 

2014-09-02  Steffen Neumann  <sneumann@ipb-halle.de>
	* Fix some minor stuff to make R CMD check happy



R/rc.qc.R
#' rc.qc
#'
#' summarize quality control for clustering and for quality control sample varation based on compound ($SpecAbund) and feature ($MSdata and $MSMSdata, if present)
#'
#' @param ramclustObj ramclustR object to analyze
#' @param qc.tag qc.tag character vector of length one or two.  If length is two, enter search string and factor name in $phenoData slot (i.e. c("QC", "sample.type"). If length one (i.e. "QC"), will search for this string in the 'sample.names' slot by default.  
#' @param remove.qc logical - if TRUE (default) QC injections will be removed from the returned ramclustObj (applies to $MSdata, $MSMSdata, $SpecAbund, $phenoData, as appropriate). If FALSE, QC samples remain.
#' @param npc number of Principle components to calcuate and plot
#' @param scale "pareto" by default: PCA scaling method used
#' @param outfile.basename base name of output files. Extensions added internally. default = "ramclustQC"
#' @param view.hist logical.  should histograms be plotted? 
#' @details plots a ramclustR summary plot.  first page represents the correlation of each cluster to all other clusters, sorted by retention time.  large blocks of yellow along the diaganol indicate either poor clustering or a group of coregulated metabolites with similar retention time.  It is an imperfect diagnostic, partuclarly with lipids on reverse phase LC or sugars on HILIC LC systems.  Page 2: histogram of r values from page 1 - only r values one position from the diagonal are used.  Pages 3:5 - PCA results, with QC samples colored red.  relative standard deviation calculated as sd(QC PC scores) / sd(all PC scores).  Page 6: histogram of CV values for each compound int he dataset, QC samples only.  
#' @return   new RC object. Saves output summary plots to pdf and .csv summary tables to new 'QC' directory. If remove.qc = TRUE, moves QC samples to new $QC slot from original position. 
#' @references Broeckling CD, Afsar FA, Neumann S, Ben-Hur A, Prenni JE. RAMClust: a novel feature clustering method enables spectral-matching-based annotation for metabolomics data. Anal Chem. 2014 Jul 15;86(14):6812-7. doi: 10.1021/ac501530d.  Epub 2014 Jun 26. PubMed PMID: 24927477.
#' @references Broeckling CD, Ganna A, Layer M, Brown K, Sutton B, Ingelsson E, Peers G, Prenni JE. Enabling Efficient and Confident Annotation of LC-MS Metabolomics Data through MS1 Spectrum and Time Prediction. Anal Chem. 2016 Sep 20;88(18):9226-34. doi: 10.1021/acs.analchem.6b02479. Epub 2016 Sep 8. PubMed PMID: 7560453.
#' @concept ramclustR
#' @concept RAMClustR
#' @concept metabolomics
#' @concept mass spectrometry
#' @concept clustering
#' @concept feature
#' @concept MSFinder
#' @concept xcms
#' @author Corey Broeckling
#' @export

rc.qc<-function(ramclustObj=NULL,
                qc.tag="QC",
                remove.qc = FALSE,
                npc=4,
                scale="pareto",
                outfile.basename ="ramclustQC",
                view.hist = TRUE
                
){
  
  if(is.null(ramclustObj)) {
    stop("must supply ramclustObj as input.  i.e. ramclustObj = RC", '\n')
  }
  
  if(is.null(qc.tag)) {
    stop("qc.tag = NULL; qc.tag must be defined to enable QC variance examination.", '\n')
  }
  
  if(is.null(outfile.basename)) {
    outfile.basename <- "ramclustQC"
  }
  
  do.sets <- c("MSdata", "SpecAbund")
  
  if(is.null(ramclustObj$SpecAbund)) {
    do.sets <- do.sets[!(do.sets %in% "SpecAbund")]
  } 
  
  do.sets.rows <- sapply(
    c(do.sets, "phenoData"), 
    FUN = function(x) {
      nrow(ramclustObj[[x]])
    })
  
  if(!sd(do.sets.rows) == 0) {
    stop("number of rows in MSdata, SpecAbund, and phenoData sets are not identical.")
  }
  
  ## define QC samples in each set
  if(length(qc.tag) == 1) {
    qc <- grepl(qc.tag[1], ramclustObj$phenoData$sample.names)
  } 
  if(length(qc.tag) == 2) {
    qc <- grepl(qc.tag[1], ramclustObj$phenoData[[qc.tag[2]]])
  }
  
  if(length(which(qc)) == 0) {
    stop("no QC samples found using the qc.tag ", "'", qc.tag, "'", '\n')
  }
  
  ## create directory
  dir.create("QC")
  
  ## if cv threshold for compounds has been applied, use it, else create 'all cmpds' vector
  if(!is.null(ramclustObj$SpecAbund)) {
    if(!is.null(ramclustObj$cmpd.use)) {
      cmpd.use <- ramclustObj$cmpd.use
    } else {
      cmpd.use <- rep(TRUE, length(ramclustObj$ann))
    }
  }
  
  #visualize clustering
  ## if clustering was perfect, we should see a normal distribution of 
  ## correlational r values 1 step from the diagonal
  ## imperfect clustering introduces right skew
  ## load("datasets/RCobject.Rdata")
  if(!is.null(ramclustObj$clrt)) {
    
    ## create file to collect figures. 
    pdf(file=paste("QC/", "ramclust_clustering_diagnostic.pdf", sep=""), 
        useDingbats=FALSE, width=8, height=8)  
    o<-order(ramclustObj$clrt[cmpd.use])
    c<-cor(ramclustObj$SpecAbund[,cmpd.use][,o])
    d<-diag(as.matrix((c[2:(nrow(c)), 1:ncol(c)-1])))
    hist(d, breaks=50, main="")
    title(main="histogram of pearson's r for each cluster to its adjacent cluster (by time)", cex.main=0.8,
          sub=paste("skew =", round(e1071::skewness(d), digits=3), " :values near zero are better", '\n', 
                    'WARNING:metabolic relationships will confound interpretation of this plot'), cex.sub=0.6)
    
    # ideally heatmap will have a bright yellow diagonal with no yellow squares near the diagonal
    # this is slow for larger numbers of clusters
    gplots::heatmap.2(c^2, trace="none", dendrogram="none", Rowv=FALSE, Colv=FALSE, main="pearsons r^2, clusters sorted by rt", cex.main=0.5,
                      cexRow=0.02 + 1/log10(length(o)), cexCol=0.02 + 1/log10(length(o)))
    dev.off()
  }
  
  
  ## PCA of QC samples
  ## histogram of feature and/or compound CVs for QC samples
  
  qc <- which(qc)
  
  cols<-rep(8, nrow(ramclustObj$phenoData))
  cols[qc]<-2
  
  for(x in do.sets) {
    
    ## PCA plot
    if(x == "SpecAbund") {
      td <- ramclustObj[[x]][,cmpd.use]
    } else {
      td <- ramclustObj[[x]]
    }
    
    # if(!is.null(ramclustObj$MSMSdata) & x == "MSdata") {
    #   td <- td + ramclustObj$MSMSdata
    # }
    
    if(min(dim(td)) < npc) {npc <- min(dim(td))}
    PCA<-pcaMethods::pca(td, scale=scale, nPcs=npc, center=TRUE)
    sc<-PCA@scores
    write.csv(sc, file = paste0("QC/", outfile.basename, "_", x, "_pcascores.csv"))
    pdf(file = paste0("QC/", outfile.basename, "_", x, "_qc_diagnostic.pdf"), useDingbats=FALSE, width=8, height=8)  
    
    ld<-PCA@loadings
    for(i in 1:(ncol(sc)-1)) {
      plot(sc[,i], sc[,i+1], col=cols, pch=19, main=paste(
        "PCA analysis:", x, if(x == "SpecAbund") {
          "(compounds)"
        } else {"(features)"}
      ),
      xlab=paste("PC", i, "::  r^2 =", round(PCA@R2[i], digits=2), "::  QC(rel sd) = ", 
                 round(sd(sc[qc,i])/sd(sc[,i]), digits=2) ),
      ylab=paste("PC", i+1, "::  r^2 =", round(PCA@R2[i+1], digits=2), "::  QC(rel sd) = ", 
                 round(sd(sc[qc,i+1])/sd(sc[,i+1]), digits=2) )
      )
      legend(qc.tag, text.col=2, x="topright", bty="n")
    }
    
    ## histogram of QC relative standard deviations for all compounds/clusters
    sds<-apply(td[qc,], 2, FUN="sd", na.rm=TRUE)
    #cat(sds, '\n')
    means<-apply(td[qc,], 2, FUN="mean", na.rm=TRUE)
    cvs<-sds/means
    
    if(x == "MSdata") {
      ramclustObj$qc.cv.feature <- cvs
      ramclustObj$qc.cv.feature.msdata <- cvs
      if(!is.null(ramclustObj$MSMSdata)) {
        sds<-apply(ramclustObj$MSMSdata[qc,], 2, FUN="sd", na.rm=TRUE)
        #cat(sds, '\n')
        means<-apply(ramclustObj$MSMSdata[qc,], 2, FUN="mean", na.rm=TRUE)
        msms.cvs<-sds/means
        ramclustObj$qc.cv.feature.msmsdata <- msms.cvs
        cvs <- pmin(ramclustObj$qc.cv.feature.msdata, msms.cvs)
        ramclustObj$qc.cv.feature <- cvs
      }
      
    } else {
      ramclustObj$qc.cv.cmpd <- cvs
    }
    qs<-quantile(cvs, probs=seq(0,1,0.2), na.rm=TRUE)
    hist(cvs, breaks=50, main="")
    title(paste("histogram of", x,  "CVs from QC samples"), line=2.7)
    title("20% quantiles in red on top axis", col.main =2, cex.main=0.7, line=2)
    axis(side=3, col=2, col.ticks=2, col.axis=2, round(qs, digits=3), labels=TRUE, las=2, cex.axis=0.4)
    dev.off()
    
    if(view.hist) {
      hist(cvs, breaks=50, main="")
      title(paste("histogram of", x,  "CVs from QC samples"), line=2.7)
      title("20% quantiles in red on top axis", col.main =2, cex.main=0.7, line=2)
      axis(side=3, col=2, col.ticks=2, col.axis=2, round(qs, digits=3), labels=TRUE, las=2, cex.axis=0.4)
      
    }
    
    
    if(x == "SpecAbund") {
      out <- data.frame(
        "cmpd" = ramclustObj$cmpd[cmpd.use],
        "annotation" = ramclustObj$ann[cmpd.use],
        "rt" = ramclustObj$clrt[cmpd.use], 
        "rdsd" = ramclustObj$clrtsd[cmpd.use],
        "mean.int" = means,
        "cv" = cvs
      )
    } else {
      out <- data.frame(
        "mz" = ramclustObj$fmz,
        "rt" = ramclustObj$frt,
        "mean.int" = means,
        "cv" = cvs
      )
      if(length(ramclustObj$labels) > 0) {
        out <- data.frame(
          out, 
          "feature" = ramclustObj$labels,
          "cluster" = ramclustObj$featclus
        )
      }
    }
    write.csv(out, file = paste0("QC/", outfile.basename, "_", x, "_cv_summar.csv"))
  }
  
  if(remove.qc) {
    ramclustObj$qc <- list()
    for(x in c("phenoData", do.sets)) {
      ramclustObj$qc[[x]] <- ramclustObj[[x]][qc,]
      ramclustObj[[x]] <- ramclustObj[[x]][-qc,]
    }
  }
  
  ramclustObj$history$qc.summary <- paste(
    
    "Variance in quality control samples was described using the",
    "rc.qc function within ramclustR. Summary statistics are provided",
    "including the relative standard deviation of QC samples to all",
    "samples in PCA space, as well as the relative standard deviation",
    "of each feature/compound in QC samples, plotted as a histogram.", 
    if(!is.null(ramclustObj$cmpd.use)) {" Only compounds which passed the CV filter are reported."}
  )
  
  return(ramclustObj)
}


